File: service/lib/dinstaller/dbus/clients/storage.rb
# frozen_string_literal: true

# Copyright (c) [2022] SUSE LLC
#
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of version 2 of the GNU General Public License as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
# more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, contact SUSE LLC.
#
# To contact SUSE LLC about this file by physical or electronic mail, you may
# find current contact information at www.suse.com.

require "dinstaller/dbus/clients/base"
require "dinstaller/dbus/clients/with_service_status"
require "dinstaller/dbus/clients/with_progress"
require "dinstaller/dbus/clients/with_validation"

module DInstaller
  module DBus
    module Clients
      # D-Bus client for storage configuration
      class Storage < Base
        include WithServiceStatus
        include WithProgress
        include WithValidation

        PROPOSAL_IFACE = "org.opensuse.DInstaller.Storage.Proposal1"
        private_constant :PROPOSAL_IFACE

        def initialize
          super

          @dbus_object = service["/org/opensuse/DInstaller/Storage1"]
          @dbus_object.introspect

          @dbus_proposal = service.object("/org/opensuse/DInstaller/Storage/Proposal1")
          @dbus_proposal.introspect
        end

        # @return [String]
        def service_name
          @service_name ||= "org.opensuse.DInstaller.Storage"
        end

        # Starts the probing process
        #
        # If a block is given, the method returns immediately and the probing is performed in an
        # asynchronous way.
        #
        # @param done [Proc] Block to execute once the probing is done
        def probe(&done)
          dbus_object.Probe(&done)
        end

        # Performs the packages installation
        def install
          dbus_object.Install
        end

        # Cleans-up the storage stuff after installation
        def finish
          dbus_object.Finish
        end

        # Devices available for the installation
        #
        # @return [Array<String>] name of the devices
        def available_devices
          dbus_proposal[PROPOSAL_IFACE]["AvailableDevices"]
            .map(&:first)
        end

        # Devices selected for the installation
        #
        # @return [Array<String>] name of the devices
        def candidate_devices
          dbus_proposal[PROPOSAL_IFACE]["CandidateDevices"]
        end

        # Actions to perform in the storage devices
        #
        # @return [Array<String>]
        def actions
          dbus_proposal[PROPOSAL_IFACE]["Actions"].map do |a|
            a["Text"]
          end
        end

        # Calculates the storage proposal with the given devices
        #
        # @param candidate_devices [Array<String>] name of the new candidate devices
        def calculate(candidate_devices)
          dbus_proposal.Calculate({ "CandidateDevices" => candidate_devices })
        end

      private

        # @return [::DBus::Object]
        attr_reader :dbus_object

        # @return [::DBus::Object]
        attr_reader :dbus_proposal
      end
    end
  end
end

File: web/src/hooks/model/system/storage.ts
/*
 * Copyright (c) [2025] SUSE LLC
 *
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, contact SUSE LLC.
 *
 * To contact SUSE LLC about this file by physical or electronic mail, you may
 * find current contact information at www.suse.com.
 */

import { useCallback } from "react";
import { useSuspenseQuery } from "@tanstack/react-query";
import { systemQuery } from "~/hooks/model/system";
import { flatDevices, findDevices, findDeviceByName } from "~/model/system/storage";
import type { System, Storage } from "~/model/system";

const selectSystem = (data: System | null): Storage.System => data?.storage;

function useSystem(): Storage.System | null {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectSystem,
  });
  return data;
}

const selectEncryptionMethods = (data: System | null): Storage.EncryptionMethod[] =>
  data?.storage?.encryptionMethods || [];

function useEncryptionMethods(): Storage.EncryptionMethod[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectEncryptionMethods,
  });
  return data;
}

const enum DeviceGroup {
  AvailableDrives = "availableDrives",
  CandidateDrives = "candidateDrives",
  AvailableMdRaids = "availableMdRaids",
  CandidateMdRaids = "candidateMdRaids",
}

function selectDeviceGroups(data: System | null, groups: DeviceGroup[]): Storage.Device[] {
  if (!data?.storage) return [];
  const sids = groups.flatMap((g) => data.storage[g]);
  return findDevices(data.storage, sids);
}

const selectAvailableDrives = (data: System | null): Storage.Device[] =>
  selectDeviceGroups(data, [DeviceGroup.AvailableDrives]);

/**
 * Hook that returns the list of available drives for installation.
 */
function useAvailableDrives(): Storage.Device[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectAvailableDrives,
  });
  return data;
}

const selectCandidateDrives = (data: System | null): Storage.Device[] =>
  selectDeviceGroups(data, [DeviceGroup.CandidateDrives]);

/**
 * Hook that returns the list of candidate drives for installation.
 */
function useCandidateDrives(): Storage.Device[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectCandidateDrives,
  });
  return data;
}

const selectAvailableMdRaids = (data: System | null): Storage.Device[] =>
  selectDeviceGroups(data, [DeviceGroup.AvailableMdRaids]);

/**
 * Hook that returns the list of available MD RAIDs for installation.
 */
function useAvailableMdRaids(): Storage.Device[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectAvailableMdRaids,
  });
  return data;
}

const selectCandidateMdRaids = (data: System | null): Storage.Device[] =>
  selectDeviceGroups(data, [DeviceGroup.CandidateMdRaids]);

/**
 * Hook that returns the list of available MD RAIDs for installation.
 */
function useCandidateMdRaids(): Storage.Device[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectCandidateMdRaids,
  });
  return data;
}

const selectAvailableDevices = (data: System | null): Storage.Device[] =>
  selectDeviceGroups(data, [DeviceGroup.AvailableDrives, DeviceGroup.AvailableMdRaids]);

/**
 * Hook that returns the list of available devices for installation.
 */
function useAvailableDevices(): Storage.Device[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectAvailableDevices,
  });
  return data;
}

const selectCandidateDevices = (data: System | null): Storage.Device[] =>
  selectDeviceGroups(data, [DeviceGroup.CandidateDrives, DeviceGroup.CandidateMdRaids]);

/**
 * Hook that returns the list of candidate devices for installation.
 */
function useCandidateDevices(): Storage.Device[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectCandidateDevices,
  });
  return data;
}

const selectDevices = (data: System | null): Storage.Device[] => data?.storage?.devices || [];

function useDevices(): Storage.Device[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectDevices,
  });
  return data;
}

const selectFlattenDevices = (data: System | null): Storage.Device[] =>
  data?.storage ? flatDevices(data.storage) : [];

function useFlattenDevices(): Storage.Device[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectFlattenDevices,
  });
  return data;
}

function useDevice(name: string): Storage.Device | null {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: useCallback(
      (data: System | null): Storage.Device | null => {
        return data?.storage ? findDeviceByName(data.storage, name) : null;
      },
      [name],
    ),
  });
  return data;
}

const selectVolumeTemplates = (data: System | null): Storage.Volume[] =>
  data?.storage?.volumeTemplates || [];

function useVolumeTemplates(): Storage.Volume[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectVolumeTemplates,
  });
  return data;
}

const selectVolumeTemplate = (data: System | null, mountPath: string): Storage.Volume | null => {
  const volumes = data?.storage?.volumeTemplates || [];
  return volumes.find((v) => v.mountPath === mountPath) || volumes.find((v) => v.mountPath === "");
};

function useVolumeTemplate(mountPath: string): Storage.Volume | null {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: useCallback((data) => selectVolumeTemplate(data, mountPath), [mountPath]),
  });
  return data;
}

const selectIssues = (data: System | null): Storage.Issue[] => data?.storage?.issues || [];

function useIssues() {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectIssues,
  });
  return data;
}

export {
  useSystem,
  useEncryptionMethods,
  useAvailableDrives,
  useCandidateDrives,
  useAvailableMdRaids,
  useCandidateMdRaids,
  useAvailableDevices,
  useCandidateDevices,
  useDevices,
  useFlattenDevices,
  useDevice,
  useVolumeTemplates,
  useVolumeTemplate,
  useIssues,
};
